feat: add Components, Security Requirement(s) models and other improvements

smoya opened the pull request:
**Description**

Some tooling need to access `components` directly to components so the model is required. Also added `CorrelationIds` as it was needed by the `Components` model.
Examples of why this is needed:

- https://github.com/asyncapi/parser-js/blob/master/lib/iterators.js#L233
- Studio 

This PR also adds the models for `SecurityRequirement`, `MessageExample` and the collection of those, `SecurityRequirements` and `MessageExamples`.
It also includes some fixes on types.

magicmatatjahu commented on docs/v1.md:
```diff
@@ -59,6 +60,23 @@
+- messageTraits(): `MessageTraits`
+- correlationIds(): `CorrelationId[]`
+- securitySchemes(): `SecuritySchemes`
+- serverBindings(): `Bindings`
+- channelBindings(): `Bindings`
+- operationBindings(): `Bindings`
+- messageBindings(): `Bindings`
```
These methods has invalid return type. `Bindings` is a collection of bindings, but for single object: `channel`, `server`, `operation`, `message` - `Bindings` takes object like https://github.com/asyncapi/spec/blob/master/spec/asyncapi.md#server-bindings-object and transform it to the list of bindings with `protocol` as `id`. For these fields spec has signature as 

```
Map[string, [Server Bindings Object | Reference Object]`
```

so we need to change it to the:

```suggestion
- serverBindings(): `Record<string, Bindings>`
- channelBindings(): `Record<string, Bindings>`
- operationBindings(): `Record<string, Bindings>`
- messageBindings(): `Record<string, Bindings>`
```

smoya replied:
Good catch. I'm going to use `Map[string, Bindings]` per convention with what we decided originally on this API to use for maps.

magicmatatjahu commented on docs/v1.md:
```diff
@@ -59,6 +60,23 @@
+- serverVariables(): `ServerVariables`
+- operationTraits(): `OperationTraits`
+- messageTraits(): `MessageTraits`
+- correlationIds(): `CorrelationId[]`
```
Should we create `CorrelationIds` model?

smoya replied:
I did at the very beginning, but then removed. The reason is that there is no collection nor map, but always an array, since correlationId doesn't have an Id to be used.

magicmatatjahu replied:
But in this case you will have key of map.

smoya replied:
Oh thats right. In components we do. 👍 adding

smoya replied:
done

magicmatatjahu reviewed (commented):
Last comment 😄

magicmatatjahu reviewed (dismissed):
LGTM!

fmvilas reviewed (approved):
LGTM 👍

smoya commented:
/rtm

smoya commented:
@fmvilas @magicmatatjahu I added the `operations()` method to `Components` as I forgot to add it.

magicmatatjahu commented on docs/v1.md:
```diff
@@ -59,6 +60,24 @@
+- servers(): `Servers`
+- channels(): `Channels`
+- messages(): `Messages`
+- operations() : `Operations`
```
Do we have agree that we will have that in v3?

smoya replied:
To have what? This is an API, nothing related with what the spec has or doesn't. 
The same we have `root.components()` we have it here under `components`.

smoya commented:
I decided to include more things into this PR so we do the review at once.

I just added:

> models for SecurityRequirement and the collection of those, SecurityRequirements.
> It also includes minor fixes on types.

cc @magicmatatjahu @jonaslagoni

magicmatatjahu commented on docs/v1.md:
```diff
@@ -226,7 +251,7 @@
 - summary() : `string` | `undefined`
 - tags() : `Tags`
-- security() : `Array<Record<string, { schema: SecurityScheme, scopes: Array<string> }>>`
+- security() : `SecurityRequirements`
```
I have problem here. `SecurityRequirements` in this PR is a list/array of singles `SecurityRequirement` which has only two methods `scheme` and `scopes` and here is a problem. In the spec `Sec` is a map of pointed schemas and scopes -> https://github.com/asyncapi/spec/blob/master/spec/asyncapi.md#security-requirement-object so you can have multiple schemas in single security "item", like this one:

```yaml
security:
  - apiKeyScheme: []
     oauthScheme: ['...']
  - openIDScheme: []
``` 

Above example means that you wanna be send/receive something to/given channel you need be authorize against `apiKeyScheme` and `oauthScheme` or only `openIDScheme`. I don't know how we should change it, to the `Array<SecurityRequirements>`? Wdyt?

magicmatatjahu replied:
Yeah, I know, it's hard to "conduct", but in spec you have array of object (of security schemes), not array of security scheme (single security scheme). If it's an object with keys then from the point of view of the spec's syntax itself you can use 2 schemas in one requirement.

cc @fmvilas

smoya replied:
Ok, now i got it (thanks @magicmatatjahu). Taking the definition of `security` field of `servers` (for example), it says it is an array of Security Requirement. Indeed it feels weird to me, but in the meantime I will change then this to become `SecurityRequirements[]`, so we are flexible enough to support the current behavior but also any future.

smoya commented:
Just added:

> `MessageExample` and `MessageExamples`.

smoya commented:
Added `id()` and `address()` to `Channel` because for some reason it was gone :O 

cc @magicmatatjahu

smoya commented:
I have simplified `MessageTrait` and `OperationTrait` by removing some methods not backed up by any use case from https://github.com/asyncapi/shape-up-process/issues/84.

Changes: https://github.com/asyncapi/parser-api/pull/71/commits/eab44c2cc5c3fd6e8319ed1660b71121edc2907a

cc @magicmatatjahu @jonaslagoni @fmvilas

smoya commented:
I've added `extensions()` method to `Bindings` model since the spec explicitly mentions:

> This object can be extended with [Specification Extensions](https://www.asyncapi.com/docs/reference/specification/v2.4.0#specificationExtensions).

cc @jonaslagoni @magicmatatjahu

smoya commented:
I've done a cleanup on some models, removing unnecessary methods, and fixing return types (especially adding missing `undefined`). The commit: https://github.com/asyncapi/parser-api/pull/71/commits/1605ee15d40e6b57140f0022e906639e0de56ee8

cc @jonaslagoni @magicmatatjahu

smoya commented:
I've added `filterByInUse()` and `filterByNotInUse()` to `Schemas`, `Channels`, `Operations`, `Messages`, and `Servers` collections.

- `filterByInUse()` returns all elements that are being used in the current document. I.e. a Channel under `document.channels`.
- `filterByNotInUse()` returns all elements that are not being used in the current document. I.e. a channel that is only declared in components but not referenced ever.

cc @jonaslagoni @magicmatatjahu @fmvilas @derberg

magicmatatjahu commented:
@smoya maybe `filterUsed` and `filterNotUsed` will be better?

smoya commented:
> @smoya maybe `filterUsed` and `filterNotUsed` will be better?

I wanted to be consistent with the name of the rest of the filters in the API. All of them have the prefix `filterBy`. We can break that if you think it is really necessary.

fmvilas commented on docs/v1.md:
```diff
@@ -1,7 +1,8 @@
 
 ## AsyncAPIDocument
 - channels() : `Channels`
+- components() : `Components`
```
I'm not 100% sure this is a good idea on an intent-driven API. If we forget about AsyncAPI, what does `components` mean? Same probably goes for `extensions` 🤔

smoya replied:
That's a really good point. And related to the discussion in https://github.com/asyncapi/parser-js/pull/612/files#r971793485. 
`components()` use case is only internal, for the schemas iterator on `parser-js` (See [code here](https://github.com/asyncapi/parser-js/blob/next-major/src/iterator.ts#L62)).

I think we deviated a bit from the Intent-Driven approach with this and went direct to mimic current `parser-js` behavior just to provide same functionalities (assigning anonymous names when missing). 
I agree `components()` doesn't represent any intent and there are no more use cases found than the one mentioned above. 

If we apply what you @fmvilas suggested in https://github.com/asyncapi/parser-api/pull/71/files#r972687041, and then we modify the iterator a bit (replacing `components` by those new statuses in the `SchemaTypesToIterate`)  we could keep the functionality of the iterator and get rid of the `components()` method in favor of those filters on each collection.

Regarding `extensions()`, I see it should be kept. How would you grab extensions from the root doc then?

cc @magicmatatjahu

magicmatatjahu replied:
Components are needed for the Studio and React Component, not for iterator. Iterator is internal functionality.

magicmatatjahu replied:
In Studio we render navigation on the left, atm channels, servers, messages and schemas, these two last point to the `components.messages` and `components.schemas` objects.

smoya replied:
> these two last point to the `components.messages` and `components.schemas` objects

Or in another way, we are showing Messages and Schemas that do have an ID (all except those declared under the operation).
Even though this doesn't seem ok to me (I think we should also show the ones with anonymous ID, but not the place to discuss), you can still show them by either filter them by their name (remove those with anonymous ID) or similar.

magicmatatjahu replied:
From private discussion:

> how you will retrieve correlationsIds and other things from components, without components() ? Will you handle raw JSON from AsyncAPI or write function in AsyncAPiDocument  like we do for schemas , messages  etc? We need components data for FormBased AsyncAPI creator and also for Drag&Drop Builder.

If even from the point of view of templates a `components()` function is not needed, then for Studio and Html-Template it is (and similar tools) is, even for bundler or optimizer (we can use new parser to handle some things for which we have to write custom logic in mentioned packages).

smoya replied:
The solution for that would be to move all the methods from `Components` to the root doc `AsyncAPI`. Some of them are already there, like `channels()`, but the following not:

- channelParameters(): `ChannelParameters`
- serverVariables(): `ServerVariables`
- operationTraits(): `OperationTraits`
- messageTraits(): `MessageTraits`
- correlationIds(): `CorrelationIds`
- securitySchemes(): `SecuritySchemes`
- serverBindings(): `Map[string, Bindings]`
- channelBindings(): `Map[string, Bindings]`
- operationBindings(): `Map[string, Bindings]`
- messageBindings(): `Map[string, Bindings]`

TBH, I now see it makes sense to move all of those to the root of the document, but let's go deep on this and find the pros and cons. 

PROS:
- Decoupling the API from what the document looks like. This is one of the main purposes of this new API.
- Related to the previous one; no breaking change if we ever remove a component, let's say Operations. Not very probable but could happen.
- If we forget about how a document looks like, it feels more natural (ergo better UX) to call `doc.correlationIds()` than `doc.components().correlationIds()`.
- As a remember, users can still use the same objects that were previously located under `Components` object though those methods.

CONS:
- Use cases where, for some reason, **all components** need to be accessed at once, like printing/showing all components in the Studio nav bar will now need to call all those methods individually where, maybe, previously the call was done by just calling **all** methods within the object (`Components`).
- Sure there are more CONS, so I need your help.

Can you folks help me finding more cons or to add weight enough to the one I mentioned? 

cc @fmvilas @magicmatatjahu @jonaslagoni

magicmatatjahu replied:
@smoya I have one problem, how you will handle extensions in `Components` object if components will be removed? `extensions()` in root will return them? In this case methods like `filterInUse` has don't any sense, because extensions are strictly related to given object not to "collections" of objects, so `extensions()` in root should return only extensions from root, not from root + components. For me, removing components does not make sense at all and I would leave this model, but I look from pov of the Studio and the tools/ideas for them that I know of, which components they need.

fmvilas replied:
What about a `componentsExtensions()` method? If the only method inside `components` is going to be `extensions()` then just merge them into one.

I like the new structure @smoya proposed. For the downside "...all components need to be accessed at once, like printing/showing all components in the Studio nav bar will now need to call all those methods individually...", I don't think it's really a problem in the case of Studio. Those components inside the sidebar should be individual components and therefore they should only grab the info they need, not the whole components section.

magicmatatjahu replied:
Calling all these methods (from components perspective) will be needed anyway, the only significant difference between the Components model and the methods at the level of the AsyncAPIDocument model is that you will need to filter these objects whether they are in components or not - navigation in Postman/Stoplight displays if given object(s) is/are in components or not and you can for example remove it/add new or go to given line in components (and that's I have in mind when it comes to Studio), so if we wanna go with method on root, I opt to having also `filterByInComponents()` method whether they are referenced in document or not.

fmvilas replied:
Agreed. But like I said in the comment below, let's not tie things to location (components) but to meaning (optional). So yeah, `filterByOptional()` should give you everything that's in components.

fmvilas commented on docs/v1.md:
```diff
@@ -20,29 +21,35 @@
+- filterBy(filter : `(channel: Channel) => boolean`) : `Channel[]`
+- filterBySend() : `Channel[]`
+- filterByReceive() : `Channel[]`
+- filterByInUse() : `Channel[]`
+- filterByNotInUse() : `Channel[]`
```
In v3, so far, a channel may be:

1. Defined but not used.
2. Defined and used by an operation.
3. Defined under components and not used.
4. Defined under components and used. (which is weird, I have to review it again)

The fact that something is under the `components` keyword, [makes it optional](https://github.com/asyncapi/spec/pull/803/files) by default.

What about we provide yet another method like `filterByOptional`? And its variations, `filterByOptionalInUse` and `filterByOptionalNotInUse` (or other names).

This would also remove the need of having a `components().channels()` function. And we can probably remove the need of all the components objects like this. What do you think?

magicmatatjahu replied:
`InUse` and `NotInUse` determines which objects are actively (defined, referenced at least once outside of components, but still to components can belong), so the 3rd case will be `NotInUse`, and 1,2 and 4 to `InUse`.

fmvilas replied:
Yeah, I get it. I mean that to get those that are defined inside `components` without explicitly mentioning `components`, we could add the `filterByOptional` method. This will return those in the `components` section. Note how I'm referring to those by their meaning (optional) and not their location (components). This is not about `InUse` or `NotInUse`, it's about optional or not optional. An optional element (those inside components) may or may not be "in use".

magicmatatjahu replied:
Ok, I also understand your pov, thanks! However next to the optional I think we should still have `inUse` and `notInUse` and by concatenating `filterByOptional` (or by `filterByInComponents` - maybe it's more verbose method)  we can execute this: `asyncapi.messages().filterByOptional().filterByInUse()` - you have all methods which are defined in components and are used in some places in channels/operations. WDYT?

fmvilas replied:
Alright, you just came to the same conclusion that I described above 😄 

> What about we provide yet another method like `filterByOptional`? And its variations, `filterByOptionalInUse` and `filterByOptionalNotInUse` (or other names).

magicmatatjahu replied:
Sorry I forgot about that first comment, I only read the second one 😆 So I agree with you. @smoya @jonaslagoni wdyt?

jonaslagoni replied:
Sounds good 👍

smoya replied:
I'm not sure if I'm super biased by current Spec already... But trying to think on the UX i find that the "optional" concept sounds hard to understand at a first glance.
I think is because we mix what they (objects located under components) mean from the point of view of the spec (optional), but also from the point of view of the application (in use/not in use).

After giving another though, I think the word `component` is being misused here: It is not a place but a [model](https://github.com/asyncapi/parser-api/#models) (not added into `parser-api` README yet).
Having a proper description of what a component is, like in https://github.com/asyncapi/spec/pull/803 is more than enough to keep using `Components` as something an AsyncAPI document will keep having probably in mid/long term. Components, those "reusable objects for different aspects of the AsyncAPI specification" that are optional by default.

My suggestion is that we keep the `root.Components()` method and just add the `components.inUse` and `components.notInUse()` methods.

If I'm missing a point, have mercy on me as it's a bit late on my place 😆

fmvilas replied:
Yeah, you're right. It's gonna make it more complicated and there's a low risk the concept of "components" disappears or gets renamed.

smoya replied:
FYI my suggestion is on a PR here https://github.com/asyncapi/parser-js/pull/619

fmvilas reviewed (approved):
I like it. Looks like this API is ready for a ride 🚀

smoya commented on docs/v1.md:
```diff
@@ -353,14 +408,25 @@
+## Servers
+- all() : `Server[]`
+- filterBy(filter : `(server: Server) => boolean`) : `Server[]`
+- filterBySend() : `Server[]`
+- filterByReceive() : `Server[]`
```
Those were based on real intents, but I can't really find it written down.

However, at a glance, i can think on the use case where a code generator is only interested in generating servers exposed by the app (local and not remotes).
I know you could get those through the channels but still is a useful shortcut.

magicmatatjahu replied:
Thanks! Good for me.

smoya commented:
I updated the API removing the `filterByInUse()` and `filterByNotInUse()` methods and added the `all*` methods in the root according to our discussion in https://github.com/asyncapi/parser-js/pull/619#discussion_r986616003

cc @magicmatatjahu @fmvilas @jonaslagoni

smoya commented:
I changed `schema.uid()` method to `schema.id()` to have consistency across all models.

Thanks @magicmatatjahu for the catch. 

cc @jonaslagoni @fmvilas

magicmatatjahu commented on docs/v1.md:
```diff
@@ -353,14 +400,23 @@
-- security() : `Array<Record<string, { schema: SecurityScheme, scopes: Array<string> }>>`
-- url() : `string` | `undefined`
+- security() : `SecurityRequirements[]`
+- url() : `string`
```
We missed `tags` from `2.5.0` version :) 

```suggestion
- url() : `string`
- tags() : `Tags`
```

smoya commented on docs/v1.md:
```diff
@@ -75,29 +104,36 @@
+
 ## Extension
 - id() : `string`
 - version() : `string`
```
Source: https://github.com/asyncapi/parser-js/pull/663#discussion_r1006894998

magicmatatjahu replied:
Thanks :)

fmvilas commented:
@magicmatatjahu can you please take care of this PR while Sergio is out?

magicmatatjahu commented:
@fmvilas No problem, I'm up-to-date with it :)

fmvilas commented:
What's missing here, @magicmatatjahu? Can we already merge it or are you still working on it?

smoya commented:
Is there something blocking this PR to be merged? Cc @magicmatatjahu @jonaslagoni 

Otherwise, i feel we can proceed.

smoya commented:
> > @smoya New things from 3.0.0 like server's`host`, `pathName`, operation's `reply` object etc. I will have new things in this PR [asyncapi/parser-js#708](https://github.com/asyncapi/parser-js/pull/708) but I need propagate them in this PR.
> 
> I would rather merge this one, which is being old enough and with tons of changes and then work on top of it. otherwise this PR is growing and lacking context with every day it passes.

wdyt about this @jonaslagoni @magicmatatjahu @fmvilas ?

fmvilas commented:
Whatever works best for you guys. Just let's make sure to have this (and the parser) as soon as possible.

magicmatatjahu commented:
Ok I think we can merge it :) I fixed all bugs that I found. I removed some not implemented methods like `extension(name: string) => Extension` - it won't be a breaking change when we will add that method in e.g. `2.1.0`.

cc @jonaslagoni Could you accept it?

magicmatatjahu commented:
/rtm
